mpqp/tools/theoretical_simulation.py

```diff
@@ -0,0 +1,210 @@
+
+    Args:
+        circuit: The circuit.
+
+    Returns:
+        The size of the trust interval (related to the Jensen-Shannon distance).
+    """
+    noiseless_circuit = QCircuit(
+        [inst for inst in circuit.instructions if not isinstance(inst, NoiseModel)]
+    )
+    noiseless_probs = theoretical_probs(noiseless_circuit)
+    noisy_probs = theoretical_probs(circuit)
+    return dist_alpha_matching(float(jensenshannon(noiseless_probs, noisy_probs)))
+
+
+def exp_id_dist(
+    circuit: QCircuit,
+    shots: int = 1024,
+    device: AvailableDevice = AWSDevice.BRAKET_LOCAL_SIMULATOR,
+):
+    """This function computes Jensen-Shannon the distance between the non noisy
+    distribution and the noisy distribution.
+
+    Args:
+        circuit: The circuit.
+        shots: Number of shots in the basis measurement.
+        device: The device to be tested.
+
+    Returns:
+        The distance between the non noisy distribution and the noisy
+        distribution.
+    """
+    noisy_probs = theoretical_probs(circuit)
+
+    noisy_circuit = circuit.without_measurements()
+    noisy_circuit.add(BasisMeasure(shots=shots))
+    mpqp_counts = _run_single(noisy_circuit, device, {}).counts
+
+    return float(jensenshannon(mpqp_counts, noisy_probs * sum(mpqp_counts)))
+
+
+def validate_noisy_circuit(
+    circuit: QCircuit,
+    shots: int = 1024,
+    device: AvailableDevice = AWSDevice.BRAKET_LOCAL_SIMULATOR,
+) -> bool:
+    """Validates our noise pipeline for a circuit.
+
+    Args:
+        circuit: The circuit (with potential noises).
+        shots: Number of shots in the basis measurement.
+        device: The device to be tested.
+
+    Returns:
+        Weather our noise pipeline matches the theory or not.
+    """
+    return exp_id_dist(circuit, shots, device) <= trust_int(circuit)
+
+
+def exp_id_dist_excess(circuit: QCircuit, shots: int = 1024) -> float:
+    """Computes the gap between theory and our noise pipeline for a circuit.
+
+    Args:
+        circuit: The circuit (with potential noises).
+        shots: Number of shots in the basis measurement.
+
+    Returns:
+        The gap between:
+
+        1. the distance between theory and our results;
+        2. and the trust interval.
+    """
+    return exp_id_dist(circuit, shots) - trust_int(circuit)
+
+
+if __name__ == "__main__":
+    from mpqp.all import *
+
+    gates = [
+        H(0),
+        X(1),
+        Y(2),
+        Z(0),
+        S(1),
+        T(0),
+        Rx(1.2324, 2),
+        Ry(-2.43, 0),
+        Rz(1.04, 1),
+        Rk(-1, 0),
+        P(-323, 2),
+        U(1.2, 2.3, 3.4, 2),
+        SWAP(2, 1),
+        CNOT(0, 1),
+        CZ(1, 2),
+    ]
+
+    probs = [0.001, 0.01, 0.1, 0.1, 0.2, 0.3]
+    xs = sum(([elt] * 6 for elt in probs), start=[])
+    shots_vals = [500, 1_000, 5_000, 10_000, 50_000, 100_000]
+    ys = shots_vals * 6
+    dists = np.array(
+        [
+            exp_id_dist_excess(QCircuit(gates + [Depolarizing(prob)]), shots)
```

Comment:
not sure we need to overthink this: it's a manual test to debug my
implementation, I could even remove it if you want since this should be formally
tested

Reply:
I think we can let it like this yeah. I mark as resolved